056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+The coming Apache Kafka 3.6 release will have such a migration process supported as GA and the ZooKeeper mode will be supported until Apache Kafka 3.7 (January 2024) but removed after that.
+Starting from Apache Kafka 4.0 (April 2024) only KRaft mode will be supported.
+
+Based on the above Apache Kafka community plan, more details in [KIP-833](https://cwiki.apache.org/confluence/display/KAFKA/KIP-833%3A+Mark+KRaft+as+Production+Ready), the Strimzi project needs to be prepared to allow users to migrate their current cluster from running in ZooKeeper mode to KRaft mode.
+
+## Proposal
+
+The ZooKeeper to KRaft migration implementation is defined by [KIP-866](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration).
+The upstream Apache Kafka community documentation also provides a [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide showing how to run the procedure manually on your cluster.
+
+One of the main points to highlights is that such a process doesn't support to migrate to a KRaft-based cluster running in "combined" mode:
+
+* Cannot migrate from a ZooKeeper based configuration to having a cluster with nodes hosting both controller and broker (so actually reducing the number of nodes by removing the ZooKeeper ones).
+* The migration works by having ZooKeeper nodes to be replaced by new KRaft controller nodes.
+* Supporting "combined" mode is actually one of the rejected [alternatives](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-CombinedModeMigrationSupport) in KIP-866.
+
+It means that the user environment should be sized to host more nodes running in parallel, because during the migration process both ZooKeeper nodes and KRaft controllers will be running.
+The end of the migration will involve shutting down the ZooKeeper nodes reducing the resources consumption.
+
+TBD: some intro about Strimzi migration process
+
+### Prerequisites
+
+An Apache Kafka cluster is running in ZooKeeper mode with `N` brokers and `M` ZooKeeper nodes, as defined by the `replicas` fields in the corresponding `Kafka` custom resource.
+The cluster has to be migrated to use `KafkaNodePool`(s) first for running the brokers (ZooKeeper nodes are never part of pools).
+Without them it’s not possible to deploy the new KRAft controllers (separated from the current Kafka brokers) as the first step of the migration.
+So the prerequisite is having the Kafka cluster still using ZooKeeper but running the Kafka brokers by using `KafkaNodePool`(s).
+The user can follow the procedure described [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) in the official documentation, in order to meet this prerequisite.
+
+Also, the minimum Kafka version should be 3.5.0 with `inter.broker.protocol.version` set to “3.5”.
```

Comment:
I will add comment when going more into the implementation details.

Reply:
@ppatierno regarding this, I just want to highlight that there is a recent PR to disables broker startup if ZK migrations are enabled while JBOD is also enabled: https://issues.apache.org/jira/browse/KAFKA-15450

Reply:
Well I am not sure we can rely on this because if we don't detect it earlier, we could apply ZK migration flag on brokers, rolling them and then? The brokers would not start up and keep crash looping? It would put the FSM in a bad state not so easy to recover. If we detect it early we deny the migration earlier.

Reply:
Yeah, just wanted to let you know this change in behavior.